README.md
# TYPO3 Roadmap Plugin for TYPO3.org

This plugin will provide records for Majorversions, Minorversions and PHP Versions.

These records will then generate a dynamic, exportable chart (useful in 
presentations) as well as a matrix of compatible PHP versions.

* Make sure to set the estimated releases in each minor version.
* Make sure to set the correct support times in each major version.


ChangeLog
2016-02-23  Mathias Schreiber  <mathias.schreiber@typo3.org>

	* Initial code generated with kickstarter


tca.php
<?php
if (!defined('TYPO3_MODE')) {
    die ('Access denied.');
}

$TCA['tx_typo3roadmap_majorversion'] = array(
    'ctrl' => $TCA['tx_typo3roadmap_majorversion']['ctrl'],
    'interface' => array(
        'showRecordFieldList' => 'hidden,starttime,endtime,title,bodytext,regularsupport,prioritysupport,extendedsupport,phpversions'
    ),
    'feInterface' => $TCA['tx_typo3roadmap_majorversion']['feInterface'],
    'columns' => array(
        'hidden' => array(
            'exclude' => 1,
            'label' => 'LLL:EXT:lang/locallang_general.xml:LGL.hidden',
            'config' => array(
                'type' => 'check',
                'default' => '0'
            )
        ),
        'starttime' => array(
            'exclude' => 1,
            'label' => 'LLL:EXT:lang/locallang_general.xml:LGL.starttime',
            'config' => array(
                'type' => 'input',
                'size' => '8',
                'max' => '20',
                'eval' => 'date',
                'default' => '0',
                'checkbox' => '0'
            )
        ),
        'endtime' => array(
            'exclude' => 1,
            'label' => 'LLL:EXT:lang/locallang_general.xml:LGL.endtime',
            'config' => array(
                'type' => 'input',
                'size' => '8',
                'max' => '20',
                'eval' => 'date',
                'checkbox' => '0',
                'default' => '0',
                'range' => array(
                    'upper' => mktime(3, 14, 7, 1, 19, 2038),
                    'lower' => mktime(0, 0, 0, date('m') - 1, date('d'), date('Y'))
                )
            )
        ),
        'title' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_majorversion.title',
            'config' => array(
                'type' => 'input',
                'size' => '30',
                'eval' => 'required,trim',
            )
        ),
        'bodytext' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_majorversion.bodytext',
            'config' => array(
                'type' => 'text',
                'cols' => '30',
                'rows' => '5',
                'wizards' => array(
                    '_PADDING' => 2,
                    'RTE' => array(
                        'notNewRecords' => 1,
                        'RTEonly' => 1,
                        'type' => 'script',
                        'title' => 'Full screen Rich Text Editing|Formatteret redigering i hele vinduet',
                        'icon' => 'wizard_rte2.gif',
                        'script' => 'wizard_rte.php',
                    ),
                ),
            )
        ),
        'developmentstart' => array(
            'exclude' => 0,
            'label' => 'Begin of Development',
            'config' => array(
                'type' => 'input',
                'size' => '8',
                'max' => '20',
                'eval' => 'date',
                'checkbox' => '0',
                'default' => '0'
            )
        ),
        'regularsupport' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_majorversion.regularsupport',
            'config' => array(
                'type' => 'input',
                'size' => '8',
                'max' => '20',
                'eval' => 'date',
                'checkbox' => '0',
                'default' => '0'
            )
        ),
        'prioritysupport' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_majorversion.prioritysupport',
            'config' => array(
                'type' => 'input',
                'size' => '8',
                'max' => '20',
                'eval' => 'date',
                'checkbox' => '0',
                'default' => '0'
            )
        ),
        'extendedsupport' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_majorversion.extendedsupport',
            'config' => array(
                'type' => 'input',
                'size' => '8',
                'max' => '20',
                'eval' => 'date',
                'checkbox' => '0',
                'default' => '0'
            )
        ),
        'phpversions' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_majorversion.phpversions',
            'config' => array(
                'type' => 'select',
                'foreign_table' => 'tx_typo3roadmap_phpversion',
                'foreign_table_where' => 'ORDER BY tx_typo3roadmap_phpversion.uid',
                'size' => 10,
                'minitems' => 0,
                'maxitems' => 10,
                "MM" => "tx_typo3roadmap_majorversion_phpversions_mm",
            )
        ),
        'minorversions' => array(
            'exclude' => 0,
            'label' => 'Minor versions',
            'config' => array(
                'type' => 'inline',
                'foreign_table' => 'tx_typo3roadmap_minorversion',
                'foreign_field' => 'parent'
            ),
        ),
    ),
    'types' => array(
        '0' => array(
            'showitem' => '
		hidden;;1;;1-1-1,
		title;;;;2-2-2,
		bodytext;;;richtext[]:rte_transform[mode=ts];3-3-3,
		--div--;Support Times,
		developmentstart,
		regularsupport,
		prioritysupport,
		extendedsupport,
		--div--;PHP Versions,
		phpversions,
		--div--;Minor Versions,
		minorversions
		'
        )
    ),
    'palettes' => array(
        '1' => array('showitem' => 'starttime, endtime')
    )
);


$TCA['tx_typo3roadmap_minorversion'] = array(
    'ctrl' => $TCA['tx_typo3roadmap_minorversion']['ctrl'],
    'interface' => array(
        'showRecordFieldList' => 'hidden,version,link,focus,estimated,released,parent'
    ),
    'feInterface' => $TCA['tx_typo3roadmap_minorversion']['feInterface'],
    'columns' => array(
        'hidden' => array(
            'exclude' => 1,
            'label' => 'LLL:EXT:lang/locallang_general.xml:LGL.hidden',
            'config' => array(
                'type' => 'check',
                'default' => '0'
            )
        ),
        'version' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_minorversion.version',
            'config' => array(
                'type' => 'input',
                'size' => '30',
                'eval' => 'required,trim',
            )
        ),
        'link' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_minorversion.link',
            'config' => array(
                'type' => 'input',
                'size' => '30',
                'wizards' => array(
                    '_PADDING' => 2,
                    'link' => array(
                        'type' => 'popup',
                        'title' => 'Link',
                        'icon' => 'link_popup.gif',
                        'script' => 'browse_links.php?mode=wizard',
                        'JSopenParams' => 'height=300,width=500,status=0,menubar=0,scrollbars=1'
                    ),
                ),
                'eval' => 'trim',
            )
        ),
        'focus' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_minorversion.focus',
            'config' => array(
                'type' => 'input',
                'size' => '30',
                'eval' => 'required,trim',
            )
        ),
        'estimated' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_minorversion.estimated',
            'config' => array(
                'type' => 'input',
                'size' => '8',
                'max' => '20',
                'eval' => 'date',
                'checkbox' => '0',
                'default' => '0'
            )
        ),
        'released' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_minorversion.released',
            'config' => array(
                'type' => 'input',
                'size' => '8',
                'max' => '20',
                'eval' => 'date',
                'checkbox' => '0',
                'default' => '0'
            )
        ),
        'parent' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_minorversion.parent',
            'config' => array(
                'type' => 'input',
                'size' => '4',
                'max' => '4',
                'eval' => 'int',
                'checkbox' => '0',
                'range' => array(
                    'upper' => '1000',
                    'lower' => '10'
                ),
                'default' => 0
            )
        ),
    ),
    'types' => array(
        '0' => array('showitem' => 'hidden;;1;;1-1-1, version, link, focus, estimated, released, parent')
    ),
    'palettes' => array(
        '1' => array('showitem' => '')
    )
);


$TCA['tx_typo3roadmap_phpversion'] = array(
    'ctrl' => $TCA['tx_typo3roadmap_phpversion']['ctrl'],
    'interface' => array(
        'showRecordFieldList' => 'hidden,version,activesupport,securitysupport'
    ),
    'feInterface' => $TCA['tx_typo3roadmap_phpversion']['feInterface'],
    'columns' => array(
        'hidden' => array(
            'exclude' => 1,
            'label' => 'LLL:EXT:lang/locallang_general.xml:LGL.hidden',
            'config' => array(
                'type' => 'check',
                'default' => '0'
            )
        ),
        'version' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_phpversion.version',
            'config' => array(
                'type' => 'input',
                'size' => '30',
            )
        ),
        'activesupport' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_phpversion.activesupport',
            'config' => array(
                'type' => 'input',
                'size' => '8',
                'max' => '20',
                'eval' => 'date',
                'checkbox' => '0',
                'default' => '0'
            )
        ),
        'securitysupport' => array(
            'exclude' => 0,
            'label' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_phpversion.securitysupport',
            'config' => array(
                'type' => 'input',
                'size' => '8',
                'max' => '20',
                'eval' => 'date',
                'checkbox' => '0',
                'default' => '0'
            )
        ),
    ),
    'types' => array(
        '0' => array('showitem' => 'hidden;;1;;1-1-1, version, activesupport, securitysupport')
    ),
    'palettes' => array(
        '1' => array('showitem' => '')
    )
);

ext_tables.php
<?php
if (!defined('TYPO3_MODE')) {
    die ('Access denied.');
}
if (!isset($_EXTKEY)) {
    $_EXTKEY = 'typo3_roadmap';
}
$TCA['tx_typo3roadmap_majorversion'] = array(
    'ctrl' => array(
        'title' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_majorversion',
        'label' => 'title',
        'tstamp' => 'tstamp',
        'crdate' => 'crdate',
        'cruser_id' => 'cruser_id',
        'sortby' => 'sorting',
        'delete' => 'deleted',
        'dividers2tabs' => 1,
        'enablecolumns' => array(
            'disabled' => 'hidden',
            'starttime' => 'starttime',
            'endtime' => 'endtime',
        ),
        'dynamicConfigFile' => t3lib_extMgm::extPath($_EXTKEY) . 'tca.php',
        'iconfile' => t3lib_extMgm::extRelPath($_EXTKEY) . 'icon_tx_typo3roadmap_majorversion.gif',
    ),
);

$TCA['tx_typo3roadmap_minorversion'] = array(
    'ctrl' => array(
        'title' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_minorversion',
        'label' => 'version',
        'tstamp' => 'tstamp',
        'crdate' => 'crdate',
        'cruser_id' => 'cruser_id',
        'sortby' => 'sorting',
        'delete' => 'deleted',
        'enablecolumns' => array(
            'disabled' => 'hidden',
        ),
        'dynamicConfigFile' => t3lib_extMgm::extPath($_EXTKEY) . 'tca.php',
        'iconfile' => t3lib_extMgm::extRelPath($_EXTKEY) . 'icon_tx_typo3roadmap_minorversion.gif',
    ),
);

$TCA['tx_typo3roadmap_phpversion'] = array(
    'ctrl' => array(
        'title' => 'LLL:EXT:typo3_roadmap/locallang_db.xml:tx_typo3roadmap_phpversion',
        'label' => 'version',
        'tstamp' => 'tstamp',
        'crdate' => 'crdate',
        'cruser_id' => 'cruser_id',
        'sortby' => 'sorting',
        'delete' => 'deleted',
        'enablecolumns' => array(
            'disabled' => 'hidden',
        ),
        'dynamicConfigFile' => t3lib_extMgm::extPath($_EXTKEY) . 'tca.php',
        'iconfile' => t3lib_extMgm::extRelPath($_EXTKEY) . 'icon_tx_typo3roadmap_phpversion.gif',
    ),
);


t3lib_div::loadTCA('tt_content');
$TCA['tt_content']['types']['list']['subtypes_excludelist'][$_EXTKEY . '_pi1'] = 'layout,select_key,pages';


t3lib_extMgm::addPlugin(array(
    'LLL:EXT:typo3_roadmap/locallang_db.xml:tt_content.list_type_pi1',
    $_EXTKEY . '_pi1',
    t3lib_extMgm::extRelPath($_EXTKEY) . 'ext_icon.gif'
), 'list_type');


pi1/class.tx_typo3roadmap_pi1.php
<?php
/***************************************************************
 *  Copyright notice
 *
 *  (c) 2016 Mathias Schreiber <mathias.schreiber@typo3.org>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

require_once(PATH_tslib . 'class.tslib_pibase.php');

/**
 * Plugin 'TYPO3 Roadmap' for the 'typo3_roadmap' extension.
 *
 * @author        Mathias Schreiber <mathias.schreiber@typo3.org>
 * @package       TYPO3
 * @subpackage    tx_typo3roadmap
 */
class tx_typo3roadmap_pi1 extends tslib_pibase
{
    public $prefixId = 'tx_typo3roadmap_pi1';        // Same as class name
    public $scriptRelPath = 'pi1/class.tx_typo3roadmap_pi1.php';    // Path to this script relative to the extension dir.
    public $extKey = 'typo3_roadmap';    // The extension key.
    public $pi_checkCHash = true;

    /**
     * @var Tx_Fluid_View_StandaloneView
     */
    protected $view;

    /**
     * Color Codes for the charts
     *
     * @var array
     */
    protected $colors = array(
        'regular' => '#69A550',
        'security' => '#FF8700',
        'elts' => '#ffb767',
        'sprint' => '#e6e6e6',
        'stabilization' => '#983030'
    );

    /**
     * @var t3lib_DB
     */
    protected $db;

    /**
     * @var t3lib_PageRenderer
     */
    protected $pageRenderer;

    /**
     * The main method of the Plugin.
     *
     * @param string $content The Plugin content
     * @param array  $conf    The Plugin configuration
     *
     * @return string The content that is displayed on the website
     */
    public function main($content = '', array $conf)
    {
        $this->conf = $conf;
        $this->pi_setPiVarDefaults();
        $this->pi_loadLL();
        $this->initializeView();
        $this->db = $GLOBALS['TYPO3_DB'];
        $this->pageRenderer = $this->getFrontendController()->getPageRenderer();

        $majorVersions = $this->getMajorVersions();
        $phpVersions = $this->getAllAvailablePHPVersions();

        $this->renderCharts($majorVersions);

        $this->view->assign('majors', $majorVersions);
        $this->view->assign('ceUid', $this->cObj->data['uid']);
        $this->view->assign('phpVersions', $phpVersions);
        $this->view->assign('majorVersions', $majorVersions);

        $content .= $this->view->render();

        return $content;
    }

    /**
     * Initialize the Fluid StandAlone View and set the template
     */
    protected function initializeView()
    {
        $this->view = t3lib_div::makeInstance('Tx_Fluid_View_StandaloneView');
        $this->view->setTemplatePathAndFilename(t3lib_extMgm::extPath($this->extKey) .
            '/Resources/Private/Templates/Roadmap.html');
    }

    /**
     * Fetches all major versions from the database
     *
     * @return array|NULL
     */
    protected function getMajorVersions()
    {
        $rows = $this->db->exec_SELECTgetRows(
            '*',
            'tx_typo3roadmap_majorversion',
            '1 = 1 ' . $this->cObj->enableFields('tx_typo3roadmap_majorversion'),
            'sorting'
        );
        foreach ($rows as $index => $singleRow) {
            foreach ($singleRow as $fieldName => $fieldValue) {
                if (t3lib_div::testInt($fieldValue)) {
                    $rows[$index][$fieldName] = (int)$fieldValue;
                }
            }
            $rows[$index]['minorversions'] = $this->getMinorVersions($singleRow['uid']);
            $rows[$index]['phpVersions'] = $this->getSupportedPHPVersions($singleRow['uid']);
        }

        return $rows;
    }

    /**
     * Fetches the minor versions of a given major version
     *
     * @param int $majorVersion The major version uid to fetch the minors for
     *
     * @return array|NULL
     */
    protected function getMinorVersions($majorVersion)
    {
        $rows = $this->db->exec_SELECTgetRows(
            '*',
            'tx_typo3roadmap_minorversion',
            'parent = ' . (int)$majorVersion . '' . $this->cObj->enableFields('tx_typo3roadmap_minorversion'),
            'sorting'
        );

        return $rows;
    }

    /**
     * @param int $majorVersion
     *
     * @return array
     */
    protected function getSupportedPHPVersions($majorVersion)
    {
        $rows = $this->db->exec_SELECTgetRows(
            'uid, version',
            'tx_typo3roadmap_majorversion_phpversions_mm
  INNER JOIN tx_typo3roadmap_phpversion ON tx_typo3roadmap_phpversion.uid = tx_typo3roadmap_majorversion_phpversions_mm.uid_foreign',
            'tx_typo3roadmap_majorversion_phpversions_mm.uid_local = ' . (int)$majorVersion .
            $this->cObj->enableFields('tx_typo3roadmap_phpversion'),
            '',
            '',
            '',
            'uid'
        );

        return $rows;
    }

    /**
     * @return array|NULL
     */
    protected function getAllAvailablePHPVersions()
    {
        $rows = $this->db->exec_SELECTgetRows(
            'uid, version',
            'tx_typo3roadmap_phpversion',
            '1 = 1 ' . $this->cObj->enableFields('tx_typo3roadmap_phpversion'),
            '',
            'version ASC',
            '',
            'uid'
        );
        if ($rows === null) {
            return array();
        }

        return $rows;
    }

    /**
     * Renders the LTS charts
     *
     * @param array $majorVersions
     */
    protected function renderCharts(array $majorVersions)
    {
        $this->pageRenderer->addJsLibrary('amcharts', 'https://www.amcharts.com/lib/3/amcharts.js');
        $this->pageRenderer->addJsLibrary('amcharts_serial', 'https://www.amcharts.com/lib/3/serial.js');
        $this->pageRenderer->addJsLibrary('amcharts_gantt', 'https://www.amcharts.com/lib/3/gantt.js');
        $this->pageRenderer->addJsLibrary('amcharts_lightheme', 'https://www.amcharts.com/lib/3/themes/light.js');
        $this->pageRenderer->addJsLibrary('amcharts_export', 'https://www.amcharts.com/lib/3/plugins/export/export.js');
//        $this->pageRenderer->addCssFile('https://www.amcharts.com/lib/3/plugins/export/export.css');

        $data = $this->generateChartArray($majorVersions);
        $this->view->assign('data', $data);
        $chartJs = '
        AmCharts.useUTC = true;
        var today = new Date();
		var chart = AmCharts.makeChart("charts", {
			"type": "gantt",
			"theme": "light",
//			"marginRight": 70,
			"period": "YYYY",
			"dataDateFormat": "YYYY-MM-DD",
			"columnWidth": 0.65,
			"precision": 1,
			"valueAxis": {
				"type": "date",
				"autoGridCount": false,
				"gridCount": 24,
				"guides": [
				    {
					"value": today,
					"toValue": today,
					"lineAlpha": 1,
					"lineThickness": 1,
					"inside": true,
					"labelRotation": 90,
					"label": "Today",
					"above": true
				    }
				]
			},
			"graph": {
				"fillAlphas": 1,
				"balloonText": "<b>[[task]]</b>: [[open]]-[[value]]"
			},
			"rotate": true,
			"categoryField": "version",
			"segmentsField": "segments",
			"colorField": "color",
			"startDateField": "start",
			"endDateField": "end",
			"dataProvider": ' . json_encode($data) . ',
			"valueScrollbar": {
				"autoGridCount": true,
				"color": "#555555"
			},
			"chartCursor": {
				"cursorColor": "#55bb76",
				"valueBalloonsEnabled": false,
				"cursorAlpha": 0,
				"valueLineAlpha": 0.5,
				"valueLineBalloonEnabled": true,
				"valueLineEnabled": true,
				"zoomable": false,
				"valueZoomable": true
			},
			"export": {
				"enabled": true,
				"divId": "exportContainer",
                "position": "bottom-right",
                "fileName": "typo3-support-times",
				"menu": [
				    "PNG","PDF", "SVG"
				]
			}
		});
        ';
        $this->pageRenderer->addJsFooterInlineCode('ltsChart', $chartJs);
    }

    /**
     * Central date formatter
     *
     * @param int $timeStamp
     *
     * @return bool|string
     */
    protected function dateFromTimestamp($timeStamp = 0)
    {
        return date('Y-m-d', $timeStamp);
    }

    /**
     * Creates the raw JSON data for the chart
     * This is rather complex so take your time
     *
     * @param array $majorVersions
     *
     * @return array
     */
    protected function generateChartArray(array $majorVersions)
    {
        $data = array();
        foreach ($majorVersions as $index => $majorVersion) {
            $firstStart = (int)$majorVersion['developmentstart'];
            $data[$index]['version'] = $majorVersion['title'];
            foreach ($majorVersion['minorversions'] as $minorIndex => $minorVersion) {
                $realDate = (int)$minorVersion['estimated'];
                if ((int)$minorVersion['released'] >= (int)$minorVersion['estimated']) {
                    $realDate = (int)$minorVersion['released'];
                }

                // Check if we are in the first loop
                if ($minorIndex === 0) {
                    $sprintStart = $firstStart = (int)$majorVersion['developmentstart'];
                } else {
                    $sprintStart = $majorVersion['minorversions'][$minorIndex - 1]['estimated'];
                    if ((int)$majorVersion['minorversions'][$minorIndex - 1]['released'] >= (int)$majorVersion['minorversions'][$minorIndex - 1]['estimated']) {
                        $sprintStart = (int)$majorVersion['minorversions'][$minorIndex - 1]['released'];
                    }
                }

                //Sprint Segment
                $data[$index]['segments'][] = array(
                    'start' => $this->dateFromTimestamp($sprintStart),
                    'end' => $this->dateFromTimestamp($realDate - 1209600),
                    'color' => $this->colors['sprint'],
                    'task' => 'Sprint Phase ' . $minorVersion['version']
                );

                //Stabilization Segment
                $data[$index]['segments'][] = array(
                    'start' => $this->dateFromTimestamp($realDate - 1209600),
                    'end' => $this->dateFromTimestamp($realDate),
                    'color' => $this->colors['stabilization'],
                    'task' => 'Stabilization Phase ' . $minorVersion['version']
                );
                // Set the last release date as start of maintenance
                $firstStart = $realDate;
            }
            /**
             * Set Maintenance Times
             */
            // Regular Maintenance
            $data[$index]['segments'][] = array(
                'start' => $this->dateFromTimestamp($firstStart),
                'end' => $this->dateFromTimestamp($majorVersion['regularsupport']),
                'color' => $this->colors['regular'],
                'task' => 'Regular Maintenance'
            );
            // Security Maintenance
            $data[$index]['segments'][] = array(
                'start' => $this->dateFromTimestamp($majorVersion['regularsupport']),
                'end' => $this->dateFromTimestamp($majorVersion['prioritysupport']),
                'color' => $this->colors['security'],
                'task' => 'Priority bugfixes'
            );
            // ELTS Maintenance
            if ($majorVersion['extendedsupport'] > 0) {
                $data[$index]['segments'][] = array(
                    'start' => $this->dateFromTimestamp($majorVersion['prioritysupport']),
                    'end' => $this->dateFromTimestamp($majorVersion['extendedsupport']),
                    'color' => $this->colors['elts'],
                    'task' => 'Extended support (optional)'
                );
            }
        }

        return $data;
    }

    /**
     * @return tslib_fe
     */
    protected function getFrontendController()
    {
        return $GLOBALS['TSFE'];
    }

}

Classes/ViewHelpers/Condition/InArrayViewHelper.php
<?php

class Tx_Typo3Roadmap_ViewHelpers_Condition_InArrayViewHelper extends Tx_Fluid_Core_ViewHelper_AbstractConditionViewHelper
{
    /**
     * @param string $needle
     * @param array $array
     * @param null $arrayPath
     * @return string
     */
    public function render($needle, array $array, $arrayPath = null)
    {
        if ($this->inArray($needle, $array, $arrayPath)) {
            return $this->renderThenChild();
        } else {
            return $this->renderElseChild();
        }
    }

    /**
     * @param string $needle
     * @param array $array
     * @param $arrayPath
     * @return bool
     */
    protected function inArray($needle, array $array, $arrayPath)
    {
        foreach ($array as $key => $value) {
            if ($arrayPath !== null) {
                $arrayValue = Tx_Extbase_Reflection_ObjectAccess::getPropertyPath($value, $arrayPath);
                if ($needle === $arrayValue) {
                    return true;
                }
            } elseif ($needle === $value) {
                return true;
            }
        }

        return false;
    }
}

ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "typo3_roadmap".
 *
 * Auto generated 23-02-2016 10:46
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/
if (!isset($_EXTKEY)) {
    $_EXTKEY = 'typo3_roadmap';
}
$EM_CONF[$_EXTKEY] = array(
    'title' => 'TYPO3 Roadmap',
    'description' => 'Roadmap and Chart Generator for TYPO3.org',
    'category' => 'plugin',
    'author' => 'Mathias Schreiber',
    'author_email' => 'mathias.schreiber@typo3.org',
    'shy' => '',
    'dependencies' => '',
    'conflicts' => '',
    'priority' => '',
    'module' => '',
    'state' => 'beta',
    'internal' => '',
    'uploadfolder' => 0,
    'createDirs' => '',
    'modify_tables' => '',
    'clearCacheOnLoad' => 0,
    'lockType' => '',
    'author_company' => '',
    'version' => '0.0.0',
    'constraints' => array(
        'depends' => array(),
        'conflicts' => array(),
        'suggests' => array(),
    ),
    '_md5_values_when_last_written' => 'a:15:{s:9:"ChangeLog";s:4:"79fe";s:12:"ext_icon.gif";s:4:"1bdc";s:17:"ext_localconf.php";s:4:"d5f6";s:14:"ext_tables.php";s:4:"cf62";s:14:"ext_tables.sql";s:4:"3809";s:37:"icon_tx_typo3roadmap_majorversion.gif";s:4:"475a";s:37:"icon_tx_typo3roadmap_minorversion.gif";s:4:"a25c";s:35:"icon_tx_typo3roadmap_phpversion.gif";s:4:"401f";s:16:"locallang_db.xml";s:4:"7a43";s:10:"README.txt";s:4:"ee2d";s:7:"tca.php";s:4:"0926";s:19:"doc/wizard_form.dat";s:4:"d3cd";s:20:"doc/wizard_form.html";s:4:"2d73";s:33:"pi1/class.tx_typo3roadmap_pi1.php";s:4:"a503";s:17:"pi1/locallang.xml";s:4:"576f";}',
);
